powerdns_exporter.go

```diff
@@ -55,13 +56,55 @@ type ServerInfo struct {
 	ZonesUrl   string `json:"zones_url"`
 }
 
-// StatsEntry is used to parse JSON data from 'server/localhost/statistics' endpoint
+// StatsEntry contains the Item property which contains the actual item
 type StatsEntry struct {
+	Item interface{}
+}
+
+// StatisticItemEntry holds values of ring and map statistic items
+type StatisticItemEntry struct {
+	Name  string  `json:"name"`
+	Value float64 `json:"value,string"`
+}
+
+// StatisticItem contains just one value
+type StatisticItem struct {
 	Name  string  `json:"name"`
 	Kind  string  `json:"type"`
 	Value float64 `json:"value,string"`
 }
 
+// StatisticCollectionItem holds ring and map statistic items
+type StatisticCollectionItem struct {
+	Name  string               `json:"name"`
+	Kind  string               `json:"type"`
+	Value []StatisticItemEntry `json:"value,string"`
+}
+
+// UnmarshalJSON dynamically parses statistic items based on their type
+func (d *StatsEntry) UnmarshalJSON(data []byte) error {
+	var typ struct {
+		Kind string `json:"type"`
+	}
+
+	if err := json.Unmarshal(data, &typ); err != nil {
+		return err
+	}
+
+	switch typ.Kind {
+	case "StatisticItem":
+		d.Item = new(StatisticItem)
+	case "RingStatisticItem":
+		fallthrough
```

Comment:
Not a particular fan of fallthrough (didn't even know it existed in go :D)

Reply:
Notice that in [this implementation](https://github.com/belminf/powerdns_exporter/commit/6335b4b048643ac0ab444c694012f595d1f1c328#diff-00edf31d940d2e3bc6573969a4e964caR75) there is a distinction between Ring and Map statistic items

Ring has an extra property named `size`